types/prettier/index.d.ts

```diff
@@ -33,7 +33,7 @@ export type Doc = doc.builders.Doc;
 type ArrayElement<T> = T extends Array<infer E> ? E : never;
 
 // A union of the properties of the given object that are arrays.
-type ArrayProperties<T> = { [K in keyof T]: T[K] extends any[] ? K : never }[keyof T];
+type ArrayProperties<T> = { [K in keyof T]: NonNullable<T[K]> extends any[] ? K : never }[keyof T];
```

Comment:
Which of the following do you expect to be included?  Does it matter whether `exactOptionalPropertyTypes` is set?
```ts
interface I {
    P1: number[];
    P2?: number[];
    P3: number[] | undefined;
    P4?: number[] | undefined;
    P5: number[] | null;
    P6?: number[] | null;
}
```

Reply:
I expect all of them to be included, and exactOptionalPropertyTypes shouldn't matter.

My assumption is that at the point of use, I have a path of type AstNode<T extends I> for some T, and a node of type I (which was obtained as path.getValue()).

At this point I can say:

  if (node.P4) path.map(callback, "P4");

and it should compile. Of course, I could also leave the node.p4 check out, and it would *still* compile. But that was the case prior to #59690, and even without my fix that problem exists:

interface I { type: "I"; P1: number[]; };
interface J { type: "J"; P1: string | null; };
type Node = I | J;

Now with an AstNode<Node> I can do path.map(callback, "P1") even if I happen to be pointing at a J.

Reply:
Thanks for confirming!